credmark/types/data/contract.py

```diff
@@ -13,6 +17,18 @@
 from credmark.types.data.data_content.transparent_proxy_data import UPGRADEABLE_CONTRACT_ABI
 
 
+class Singleton:
```

Comment:
Thanks for the ideas! I have added the chain id and a check for block number. If the cache has the meta from a later block, it will not return the cache.

The cache traces are

Case 1) token.price for one block
```
[Cache] Not found in cache chain_id=1/0x1f98431c8ad98523631ae4a59f267346ea31f984/14234904
[Cache] Not found in cache chain_id=1/0x59e1f901b5c33ff6fae15b61684ebf17cca7b9b3/14234904
[Cache] Not found in cache chain_id=1/0xf7a716e2df2bde4d0ba7656c131b06b1af68513c/14234904
[Cache] Return from cache chain_id=1/0x59e1f901b5c33ff6fae15b61684ebf17cca7b9b3/14234904
[Cache] Return from cache chain_id=1/0xf7a716e2df2bde4d0ba7656c131b06b1af68513c/14234904
```

Case 2) Historical token.price
```
[Cache] Not found in cache chain_id=1/0xb4e16d0168e52d35cacd2c6185b44281ec28c9dc/14233425 <= we start from earlier blocks.
[Cache] Not found in cache chain_id=1/0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48/14233425
[Cache] Return from cache chain_id=1/0xb4e16d0168e52d35cacd2c6185b44281ec28c9dc/14233425
[Cache] Return from cache chain_id=1/0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48/14233425
...
[Cache] Return from cache chain_id=1/0xb4e16d0168e52d35cacd2c6185b44281ec28c9dc/14233634 <= returns from cache for later blocks.
[Cache] Return from cache chain_id=1/0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48/14233634
...
```

Reply:
@newtectonics, the contract.meta may need an update. It does not use the block_number now.

```
  async contractMetadataLookup(
    connection: Snowflake,
    chainId: number,
    _blockNumber: number,
    input: ContractMetadataQueryInput,
  ): Promise<ContractMetadataQueryOutput> {
    // IMPORTANT: If you update the logic to handle more parameters,
    // update the docs in the register call above.

    if (input.contractAddress) {
      return {
        contracts: await getContractMetaByAddress(
          connection,
          chainId,
          input.contractAddress,
        ),
      };
    }
...
  }
```